desktop/extensions-sdk/dev/kubernetes.md

```diff
@@ -0,0 +1,148 @@
+  "-o",
+  'custom-columns=":metadata.name"',
+  "--context",
+  "docker-desktop",
+]);
+```
+
+## Persisting the kubeconfig file
+
+Below there are different ways to persist and read the kubeconfig file from the host filesystem. Users can add/edit/remove Kubernetes context to the `kubeconfig` file at anytime.
+
+> Warning
+>
+> The `kubeconfig` file is very sensitive and if found can allow an attacker administrative access to the Kubernetes Cluster.
+{: .warning}
+
+### Extension's backend container
+
+If you need your extension to persist the `kubeconfig` file after it's been read, you can have a backend container that exposes and HTTP POST endpoint to store the content of the file either in memory or somewhere within the container filesystem. This way, if the user navigate out of the extension to another parts of Docker Desktop and then comes back, you will not need to read the `kubeconfig` file again.
+
+```typescript
+export const updateKubeconfig = async () => {
+  const kubeConfig = await ddClient.extension.host?.cli.exec("kubectl", [
+    "config",
+    "view",
+    "--raw",
+    "--minify",
+    "--context",
+    "docker-desktop",
+  ]);
+  if (kubeConfig?.stderr) {
+    console.log("error", kubeConfig?.stderr);
+    return false;
+  }
+
+  // call backend container to store the kubeconfig retrieved into the container's memory or filesystem
+  try {
+    await ddClient.extension.vm?.service?.post("/store-kube-config", {
+      data: kubeConfig?.stdout,
+    });
+  } catch (err) {
+    console.log("error", JSON.stringify(err));
+  }
+};
+```
+
+### Docker volume
+
+Volumes are the preferred mechanism for persisting data generated by and used by Docker containers. You can make use of them to persit the `kubeconfig` file.
+By persisting the `kubeconfig` in a volume you won't need to read the `kubeconfig` file again when the extension pane closes. This makes it ideal for persisting data when navigating out of the extension to another parts of Docker Desktop.
+
+```typescript
+const kubeConfig = await ddClient.extension.host?.cli.exec("kubectl", [
+  "config",
+  "view",
+  "--raw",
+  "--minify",
+  "--context",
+  "docker-desktop",
+]);
+if (kubeConfig?.stderr) {
+  console.log("error", kubeConfig?.stderr);
+  return false;
+}
+
+await ddClient.docker.cli.exec("run", [
+  "--rm",
+  "-v",
+  "my-vol:/tmp",
+  "alpine",
+  "/bin/sh",
+  "-c",
+  `"touch /tmp/.kube/config && echo '${kubeConfig?.stdout}' > /tmp/.kube/config"`,
+]);
+```
+
+### Extension's localStorage
+
+`localStorage` is one of the mechanisms of a browser's web storage. It allows users to save data as key-value pairs in the browser for later use.
+`localStorage` does not clear data when the browser (the extension pane) closes. This makes it ideal for persisting data when navigating out of the extension to another parts of Docker Desktop.
+
+```typescript
+localStorage.setItem("kubeconfig", kubeConfig);
+```
+
+```typescript
+localStorage.getItem("kubeconfig");
```

Comment:
The same way we have a warning about using localstorage [here](https://docs.docker.com/desktop/extensions-sdk/dev/oauth2-flow/#step-three-store-the-access-token), shouldn't we have one here as well?

Reply:
```md
> Warning
>
> Storing the access token in a client’s localStorage is a security risk as that storage is currently shared amongst all extensions.
{: .warning}
```

Reply:
At the moment we do not have a way to securely persist sensitive data, that's why the warning is at the beginning of the subsection: https://github.com/docker/docs/pull/15896/files#diff-93f5c7aa759604f80d718196fcedffed6913a42d1ba39e1e7c11edbd09bbfd6aR72-R75